lib/active_shipping/carriers/usps.rb

```diff
@@ -232,26 +229,44 @@ def maximum_weight
       Mass.new(70, :pounds)
     end
 
-    def extract_event_details(message)
-      return EventDetails.new unless EVENT_MESSAGE_PATTERNS.any? { |pattern| message =~ pattern }
-      description = $1.upcase
-      timestamp = $2
-      city = $3
-      state = $4
-      zip_code = $5
+    def extract_event_details(node)
+      description = node.at('Event').text.upcase
+
+      if prefix = ONLY_PREFIX_EVENTS.find { |p| description.starts_with?(p) }
+        description = prefix
+      end
+      
+      timestamp = "#{node.at('EventDate').text}, #{node.at('EventTime').text}"
+      event_code = node.at('EventCode').text
+      city = node.at('EventCity').text
+      state = node.at('EventState').text
+      zip_code = node.at('EventZIPCode').text
+
+      country_node = node.at('EventCountry')
+      country = country_node ? country_node.text : ''
+      country = 'USA' if country.empty?
 
       time = Time.parse(timestamp)
       zoneless_time = Time.utc(time.year, time.month, time.mday, time.hour, time.min, time.sec)
-      location = Location.new(city: city, state: state, postal_code: zip_code, country: 'USA')
-      EventDetails.new(description, time, zoneless_time, location)
+      location = Location.new(city: city, state: state, postal_code: zip_code, country: country)
+      EventDetails.new(description, time, zoneless_time, location, event_code)
     end
 
     protected
 
     def build_tracking_request(tracking_number, options = {})
       xml_builder = Nokogiri::XML::Builder.new do |xml|
-        xml.TrackRequest('USERID' => @options[:login]) do
-          xml.TrackID('ID' => tracking_number)
+        xml.TrackFieldRequest('USERID' => @options[:login]) do
+          xml.Revision { xml.text('1') }
+          xml.ClientIp { xml.text(@options[:client_ip] || '127.0.0.1') }
+          xml.SourceId { xml.text(@options[:source_id] || 'active_shipping') }
```

Comment:
It is undocumented what this and the client IP are really supposed to mean. They seem to do nothing. Probably for internal analytics I guess. They're mandatory.